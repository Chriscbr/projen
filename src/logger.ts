import * as chalk from 'chalk';
import { Component } from './component';
import { Project } from './project';

export class Logger extends Component {
  private _level: LogLevel;
  private usePrefix: boolean;

  constructor(project: Project) {
    super(project);

    this._level = LogLevel.INFO;
    this.usePrefix = false;
  }

  public get level() {
    return this._level;
  }

  public set level(level: LogLevel) {
    this._level = level;
  }

  public log(level: LogLevel, ...text: any[]) {
    if (level === LogLevel.OFF) {
      throw new Error('Cannot log a message with level LogLevel.OFF');
    }
    if (level <= this.level) {
      const color = this.colorForLogLevel(level);
      const prefix = this.usePrefix ? `[${this.project.name}] ` : '';
      console.error(`ðŸ¤– ${prefix}${color(...text)}`);
    }
  }

  public verbose(...text: any[]) {
    this.log(LogLevel.VERBOSE, ...text);
  }

  public debug(...text: any[]) {
    this.log(LogLevel.DEBUG, ...text);
  }

  public info(...text: any[]) {
    this.log(LogLevel.INFO, ...text);
  }

  public warn(...text: any[]) {
    this.log(LogLevel.WARN, ...text);
  }

  public error(...text: any[]) {
    this.log(LogLevel.ERROR, ...text);
  }

  private colorForLogLevel(level: LogLevel): chalk.ChalkFunction {
    switch (level) {
      case LogLevel.ERROR: return chalk.red;
      case LogLevel.WARN: return chalk.yellow;
      case LogLevel.INFO: return chalk.cyan;
      case LogLevel.DEBUG: return chalk.gray;
      case LogLevel.VERBOSE: return chalk.white;
      default:
        return chalk.white;
    }
  }
}

export enum LogLevel {
  OFF = '00.off',
  ERROR = '10.error',
  WARN = '20.warn',
  INFO = '30.info',
  DEBUG = '40.debug',
  VERBOSE = '50.verbose'
}