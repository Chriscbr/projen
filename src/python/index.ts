export * from './pip';
export * from './pytest';
export * from './python-env';
export * from './python-deps';
export * from './python-packaging';
export * from './python-project';
export * from './python-sample';
export * from './requirements-file';
export * from './venv';