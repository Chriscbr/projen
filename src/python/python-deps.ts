export interface IPythonDeps {
  /**
   * Adds a runtime dependency.
   *
   * @param spec Format `<module>@<semver>`
   */
  addDependency(spec: string): void;

  /**
   * Adds a dev dependency.
   *
   * @param spec Format `<module>@<semver>`
   */
  addDevDependency(spec: string): void;

  /**
   * Installs dependencies (called during post-synthesis).
   */
  installDependencies(): void;
}
