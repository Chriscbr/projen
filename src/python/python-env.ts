import { Task } from '../tasks';

export interface IPythonEnv {
  readonly activateTask: Task;
  readonly deactivateTask: Task;
}
