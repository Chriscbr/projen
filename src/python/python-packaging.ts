import { Task } from '../tasks';

export interface IPythonPackaging {
  /**
   * A task that packages the project for distribution.
   */
  readonly packageTask: Task;

  /**
   * A task that uploads the package to a package package repository.
   */
  readonly uploadTask: Task;
}
