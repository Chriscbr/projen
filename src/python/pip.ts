import { Component } from '../component';
import { DependencyType } from '../deps';
import { Task, TaskCategory } from '../tasks';
import { exec } from '../util';
import { IPythonDeps } from './python-deps';
import { PythonProject } from './python-project';
import { RequirementsFile } from './requirements-file';

export interface PipOptions {}

export class Pip extends Component implements IPythonDeps {
  public readonly installTask: Task;

  constructor(project: PythonProject, _options: PipOptions) {
    super(project);

    new RequirementsFile(project, 'requirements.txt', { _lazyPackages: () => this.synthDependencies() });
    new RequirementsFile(project, 'requirements-dev.txt', { _lazyPackages: () => this.synthDevDependencies() });

    this.installTask = project.addTask('install', {
      description: 'Install and upgrade dependencies',
      category: TaskCategory.BUILD,
    });
    this.installTask.exec('pip install --upgrade pip');
    this.installTask.exec('pip install -r requirements.txt');
    this.installTask.exec('pip install -r requirements-dev.txt');
  }

  private synthDependencies() {
    const dependencies: string[] = [];
    for (const pkg of this.project.deps.all) {
      if (pkg.type === DependencyType.RUNTIME) {
        dependencies.push( `${pkg.name}@${pkg.version}`);
      }
    }
    return dependencies;
  }

  private synthDevDependencies() {
    const dependencies: string[] = [];
    for (const pkg of this.project.deps.all) {
      if ([DependencyType.DEVENV].includes(pkg.type)) {
        dependencies.push( `${pkg.name}@${pkg.version}`);
      }
    }
    return dependencies;
  }

  /**
   * Adds a runtime dependency.
   *
   * @param spec Format `<module>@<semver>`
   */
  public addDependency(spec: string) {
    this.project.deps.addDependency(spec, DependencyType.RUNTIME);
  }

  /**
   * Adds a dev dependency.
   *
   * @param spec Format `<module>@<semver>`
   */
  public addDevDependency(spec: string) {
    this.project.deps.addDependency(spec, DependencyType.DEVENV);
  }

  /**
   * Installs dependencies (called during post-synthesis).
   */
  public installDependencies() {
    this.project.logger.info('Installing dependencies...');
    exec(this.installTask.toShellCommand(), { cwd: this.project.outdir });
  }
}