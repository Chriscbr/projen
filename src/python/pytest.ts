import { Component } from '../component';
import { TaskCategory } from '../tasks';
import { PythonProject } from './python-project';

export interface PytestOptions {
  /**
   * Pytest version
   *
   * @default "6.2.1"
   */
  readonly version?: string;
}

export class Pytest extends Component {
  constructor(project: PythonProject, options: PytestOptions) {
    super(project);

    const version = options.version ?? '6.2.1';

    project.depsManager?.addTestDependency(`pytest@${version}`);

    project.addTask('test', {
      description: 'Runs tests',
      category: TaskCategory.TEST,
      exec: 'pytest',
    });

    // TODO: add sample tests with `SampleDir`
  }
}