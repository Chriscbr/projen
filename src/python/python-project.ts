import { Project, ProjectOptions, ProjectType } from '../project';
import { Pip } from './pip';
import { Poetry } from './poetry';
import { Pytest, PytestOptions } from './pytest';
import { IPythonDeps } from './python-deps';
import { IPythonEnv } from './python-env';
import { IPythonPackaging } from './python-packaging';
import { PythonSample } from './python-sample';
import { Setuptools, SetuptoolsOptions } from './setuptools';
import { Venv } from './venv';


/** Allowed characters in python project names */
const PYTHON_PROJECT_NAME_REGEX = /^[A-Za-z0-9-_\.]+$/;

/**
 * Options for `PythonProject`.
 */
export interface PythonProjectOptions extends ProjectOptions {
  // -- required options --

  /**
   * Absolute path to the user's python installation.
   *
   * @default $PYTHON_PATH
   */
  readonly pythonPath: string;

  // -- general information --

  /**
   * Author's name
   */
  readonly authorName?: string;

  /**
   * Author's e-mail
   */
  readonly authorEmail?: string;

  /**
   * Manually specify package version
   */
  readonly version?: string;

  /**
   * A short project description
   */
  readonly description?: string;

  /**
   * The project license
   */
  readonly license?: string;

  /**
   * The project's homepage / website
   */
  readonly homepage?: string;

  // -- dependencies --

  /**
   * List of runtime dependencies for this project.
   *
   * Dependencies use the format: `<module>@<semver>`
   *
   * Additional dependencies can be added via `project.addDependency()`.
   *
   * @default []
   */
  readonly deps?: string[];

  /**
   * List of test dependencies for this project.
   *
   * Dependencies use the format: `<module>@<semver>`
   *
   * Additional dependencies can be added via `project.addTestDependency()`.
   *
   * @default []
   */
  readonly testDeps?: string[];

  /**
   * List of dev dependencies for this project.
   *
   * Dependencies use the format: `<module>@<semver>`
   *
   * Additional dependencies can be added via `project.addDevDependency()`.
   *
   * @default []
   */
  readonly devDeps?: string[];

  // -- core components --

  /**
   * Use pip with a requirements.txt file to track project dependencies.
   *
   * @default true
   */
  readonly pip?: boolean;

  /**
   * Use venv to manage a virtual environment for installing dependencies inside.
   *
   * @default true
   */
  readonly venv?: boolean;

  /**
   * Use setuptools with a setup.py script for packaging and distribution.
   *
   * @default - true if the project type is library
   */
  readonly setuptools?: boolean;

  /**
   * Use poetry to manage your project dependencies, virtual environment, and
   * (optional) packaging.
   *
   * @default false
   */
  readonly poetry?: boolean;

  /**
   * Setuptools options
   * @default - defaults
   */
  readonly setuptoolsOptions?: SetuptoolsOptions;

  // -- optional components --

  /**
   * Include pytest tests.
   * @default true
   */
  readonly pytest?: boolean;

  /**
   * pytest options
   * @default - defaults
   */
  readonly pytestOptions?: PytestOptions;

  /**
   * Include sample code and test if the relevant directories don't exist.
   */
  readonly sample?: boolean;
}

/**
 * Python project.
 *
 * @pjid python
 */
export class PythonProject extends Project {
  /**
   * Absolute path to the user's python installation. This will be used for
   * setting up the virtual environment.
   */
  readonly pythonPath: string;

  /**
   * Python module name (the project name, with any hyphens or periods replaced
   * with underscores).
   */
  readonly moduleName: string;

  /**
   * API for managing dependencies.
   */
  public readonly depsManager!: IPythonDeps;

  /**
   * API for mangaging the Python runtime environment.
   */
  public readonly envManager!: IPythonEnv;

  /**
   * API for managing packaging the project as a library. Only applies when the `projectType` is LIB.
   */
  public readonly packagingManager!: IPythonPackaging;

  /**
   * Pytest component.
   */
  public readonly pytest?: Pytest;

  constructor(options: PythonProjectOptions) {
    super(options);

    if (!PYTHON_PROJECT_NAME_REGEX.test(options.name)) {
      throw new Error('Python projects must only consist of alphanumeric characters, hyphens, and underscores.');
    }

    this.moduleName = this.safeName(options.name);
    this.pythonPath = options.pythonPath;

    if (options.venv ?? true) {
      this.envManager = new Venv(this, {});
    }

    if (options.pip ?? true) {
      this.depsManager = new Pip(this, {});
    }

    if (options.setuptools ?? (this.projectType === ProjectType.LIB)) {
      this.packagingManager = new Setuptools(this, {
        ...options.setuptoolsOptions,
        setupConfig: {
          authorName: options.authorName,
          authorEmail: options.authorEmail,
          version: options.version,
          description: options.description,
          license: options.license,
          homepage: options.homepage,
          ...options.setuptoolsOptions?.setupConfig,
        },
      });
    }

    // if (options.conda ?? false) {
    //   this.depsManager = new Conda(this, options);
    //   this.envManager = this.depsManager;
    // }

    // if (options.pipenv ?? false) {
    //   this.depsManager = new Pipenv(this, options);
    //   this.envManager = this.depsManager;
    // }

    if (options.poetry ?? false) {
      const poetry = new Poetry(this, options);
      this.depsManager = poetry;
      this.envManager = poetry;
      this.packagingManager = poetry;
    }

    if (!this.depsManager) {
      throw new Error('At least one tool must be chosen for managing dependencies (pip, conda, pipenv, or poetry).');
    }

    if (!this.envManager) {
      throw new Error('At least one tool must be chosen for managing the environment (venv, conda, pipenv, or poetry).');
    }

    if (!this.packagingManager) {
      if (this.projectType === ProjectType.LIB) {
        throw new Error('At least one tool must be chosen for managing packaging (setuptools or poetry).');
      } else {
        this.packagingManager = {}; // no-op packaging manager
      }
    }

    if (options.pytest ?? true) {
      this.pytest = new Pytest(this, {});
    }

    if (options.sample ?? true) {
      new PythonSample(this, {});
    }

    for (const dep of options.deps ?? []) {
      this.addDependency(dep);
    }

    for (const dep of options.devDeps ?? []) {
      this.addDevDependency(dep);
    }

    this.addDefaultGitIgnore();
  }

  /**
   * Convert an arbitrary string to a valid module filename.
   *
   * Replaces hyphens with underscores.
   *
   * @param name project name
   */
  private safeName(name: string) {
    return name.replace('-', '_').replace('.', '_');
  }

  /**
   * Adds default gitignore options for a Python project based on
   * https://github.com/github/gitignore/blob/master/Python.gitignore
   */
  private addDefaultGitIgnore() {
    this.gitignore.exclude(
      '# Byte-compiled / optimized / DLL files',
      '__pycache__/',
      '*.py[cod]',
      '*$py.class',
      '',
      '# C extensions',
      '*.so',
      '',
      '# Distribution / packaging',
      '.Python',
      'build/',
      'develop-eggs/',
      'dist/',
      'downloads/',
      'eggs/',
      '.eggs/',
      'lib/',
      'lib64/',
      'parts/',
      'sdist/',
      'var/',
      'wheels/',
      'share/python-wheels/',
      '*.egg-info/',
      '.installed.cfg',
      '*.egg',
      'MANIFEST',
      '',
      '# PyInstaller',
      '#  Usually these files are written by a python script from a template',
      '#  before PyInstaller builds the exe, so as to inject date/other infos into it.',
      '*.manifest',
      '*.spec',
      '',
      '# Installer logs',
      'pip-log.txt',
      'pip-delete-this-directory.txt',
      '',
      '# Unit test / coverage reports',
      'htmlcov/',
      '.tox/',
      '.nox/',
      '.coverage',
      '.coverage.*',
      '.cache',
      'nosetests.xml',
      'coverage.xml',
      '*.cover',
      '*.py,cover',
      '.hypothesis/',
      '.pytest_cache/',
      'cover/',
      '',
      '# Translations',
      '*.mo',
      '*.pot',
      '',
      '# Django stuff:',
      '*.log',
      'local_settings.py',
      'db.sqlite3',
      'db.sqlite3-journal',
      '',
      '# Flask stuff:',
      'instance/',
      '.webassets-cache',
      '',
      '# Scrapy stuff:',
      '.scrapy',
      '',
      '# Sphinx documentation',
      'docs/_build/',
      '',
      '# PyBuilder',
      '.pybuilder/',
      'target/',
      '',
      '# Jupyter Notebook',
      '.ipynb_checkpoints',
      '',
      '# IPython',
      'profile_default/',
      'ipython_config.py',
      '',
      '# PEP 582; used by e.g. github.com/David-OConnor/pyflow',
      '__pypackages__/',
      '',
      '# Celery stuff',
      'celerybeat-schedule',
      'celerybeat.pid',
      '',
      '# SageMath parsed files',
      '*.sage.py',
      '',
      '# Environments',
      '.env',
      '.venv',
      'env/',
      'venv/',
      'ENV/',
      'env.bak/',
      'venv.bak/',
      '',
      '# Spyder project settings',
      '.spyderproject',
      '.spyproject',
      '',
      '# Rope project settings',
      '.ropeproject',
      '',
      '# mkdocs documentation',
      '/site',
      '',
      '# mypy',
      '.mypy_cache/',
      '.dmypy.json',
      'dmypy.json',
      '',
      '# Pyre type checker',
      '.pyre/',
      '',
      '# pytype static type analyzer',
      '.pytype/',
      '',
      '# Cython debug symbols',
      'cython_debug/',
    );
  }

  /**
   * Adds a runtime dependency.
   *
   * @param spec Format `<module>@<semver>`
   */
  public addDependency(spec: string) {
    return this.depsManager.addDependency(spec);
  }

  /**
   * Adds a dev dependency.
   *
   * @param spec Format `<module>@<semver>`
   */
  public addDevDependency(spec: string) {
    return this.depsManager.addDevDependency(spec);
  }

  public postSynthesize() {
    super.postSynthesize();

    this.envManager.setupEnvironment();
    this.depsManager.installDependencies();
  }
}
