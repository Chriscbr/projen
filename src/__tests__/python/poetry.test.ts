import { LogLevel } from '../../logger';
import { PythonProject, PythonProjectOptions } from '../../python';
import { mkdtemp, synthSnapshot } from '../util';

test('poetry enabled', () => {
  const p = new TestPythonProject({
    venv: false,
    pip: false,
    setuptools: false,
    poetry: true,
    homepage: 'http://www.example.com',
    description: 'a short project description',
    license: 'Apache Software License',
  });

  expect(synthSnapshot(p)).toMatchSnapshot();
});

class TestPythonProject extends PythonProject {
  constructor(options: Partial<PythonProjectOptions> = { }) {
    super({
      ...options,
      clobber: false,
      name: 'test-python-project',
      moduleName: 'test_python_project',
      pythonPath: '/usr/bin/python',
      authorName: 'First Last',
      authorEmail: 'email@example.com',
      version: '0.1.0',
      outdir: mkdtemp(),
      logging: { level: LogLevel.OFF },
      jsiiFqn: 'projen.python.PythonProject',
    });
  }
}
