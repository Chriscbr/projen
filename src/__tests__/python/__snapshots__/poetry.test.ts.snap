// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`poetry enabled 1`] = `
Object {
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

*$py.class
*.cover
*.egg
*.egg-info/
*.log
*.manifest
*.mo
*.pot
*.py,cover
*.py[cod]
*.sage.py
*.so
*.spec
.Python
.cache
.coverage
.coverage.*
.dmypy.json
.eggs/
.env
.hypothesis/
.installed.cfg
.ipynb_checkpoints
.mypy_cache/
.nox/
.pybuilder/
.pyre/
.pytest_cache/
.pytype/
.ropeproject
.scrapy
.spyderproject
.spyproject
.tox/
.venv
.webassets-cache
/poetry.toml
/site
ENV/
MANIFEST
__pycache__/
__pypackages__/
build/
celerybeat-schedule
celerybeat.pid
cover/
coverage.xml
cython_debug/
db.sqlite3
db.sqlite3-journal
develop-eggs/
dist/
dmypy.json
docs/_build/
downloads/
eggs/
env.bak/
env/
htmlcov/
instance/
ipython_config.py
lib/
lib64/
local_settings.py
node_modules/
nosetests.xml
parts/
pip-delete-this-directory.txt
pip-log.txt
profile_default/
sdist/
share/python-wheels/
target/
var/
venv.bak/
venv/
wheels/
!/.projen/deps.json
!/.projen/tasks.json
!/pyproject.toml
",
  ".projen/deps.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "dependencies": Array [
      Object {
        "name": "pytest",
        "type": "devenv",
        "version": "6.2.1",
      },
      Object {
        "name": "python",
        "type": "runtime",
        "version": "^3.6",
      },
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "env": Object {
      "PATH": "$(echo $(poetry env info -p)/bin:$PATH)",
      "VIRTUAL_ENV": "$(poetry env info -p)",
    },
    "tasks": Object {
      "install": Object {
        "category": "00.build",
        "description": "Install and upgrade dependencies",
        "name": "install",
        "steps": Array [
          Object {
            "exec": "poetry update",
          },
        ],
      },
      "package": Object {
        "category": "20.release",
        "description": "Creates source archive and wheel for distribution.",
        "name": "package",
        "steps": Array [
          Object {
            "exec": "poetry build",
          },
        ],
      },
      "test": Object {
        "category": "10.test",
        "description": "Runs tests",
        "name": "test",
        "steps": Array [
          Object {
            "exec": "pytest",
          },
        ],
      },
      "upload": Object {
        "category": "20.release",
        "description": "Uploads the package to PyPI.",
        "name": "upload",
        "steps": Array [
          Object {
            "exec": "poetry publish",
          },
        ],
      },
      "upload:test": Object {
        "category": "20.release",
        "description": "Uploads the package against a test PyPI endpoint.",
        "name": "upload:test",
        "steps": Array [
          Object {
            "exec": "poetry publish -r testpypi",
          },
        ],
      },
    },
  },
  "README.md": "# replace this",
  "poetry.toml": "
[repositories.testpypi]
url = \\"https://test.pypi.org/legacy/\\"
",
  "pyproject.toml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

[build-system]
requires = [ \\"poetry_core>=1.0.0\\" ]
build-backend = \\"poetry.core.masonry.api\\"

[tool.poetry]
name = \\"test-python-project\\"
version = \\"0.1.0\\"
description = \\"a short project description\\"
license = \\"Apache Software License\\"
authors = [ \\"First Last <email@example.com>\\" ]
readme = \\"README.md\\"
homepage = \\"http://www.example.com\\"
scripts = { }

  [tool.poetry.dependencies]
  python = \\"^3.6\\"

  [tool.poetry.dev-dependencies]
  pytest = \\"6.2.1\\"
",
  "test_python_project/__init__.py": "__version__ = \\"0.1.0\\"
",
  "test_python_project/__main__.py": "from .example import hello

if __name__ == \\"__main__\\":
    name = input(\\"What is your name? \\")
    print(hello(name))
",
  "test_python_project/example.py": "def hello(name: str) -> str:
    \\"\\"\\"A simple greeting.
    Args:
        name (str): Name to greet.
    Returns:
        str: greeting message
    \\"\\"\\"
    return f\\"Hello {name}!\\"
",
  "tests/__init__.py": "",
  "tests/test_example.py": "import pytest

from test_python_project.example import hello

@pytest.mark.parametrize(
    (\\"name\\", \\"expected\\"),
    [
        (\\"A. Musing\\", \\"Hello A. Musing!\\"),
        (\\"traveler\\", \\"Hello traveler!\\"),
        (\\"projen developer\\", \\"Hello projen developer!\\"),
    ],
)
def test_hello(name, expected):
    \\"\\"\\"Example test with parametrization.\\"\\"\\"
    assert hello(name) == expected
",
}
`;
